### ISSUE TITLE
Model Struggling to Learn Cloth Dynamics

### ISSUE THREAD

**Issue by NathanielB123** [hbertiche/NeuralClothSim#21]
I am running training with the T-Shirt garment on around ~700 animations downloaded from Mixamo (via [mixamo_anims_downloader](https://github.com/gnuton/mixamo_anims_downloader/tree/master))

In the config, I have increased `batch_size` to 300 (any more and I exhaust the 12GB of VRAM on my RTX 4070) but have left everything else default (i.e. `temporal_window_size = 0.5` and `reflect_probability = motion_augmentation = 0.0`.

Unfortunately after training this model for 100 epochs (does not sound like much, but each is made up of 209 batches, of 300 sequences) I am struggling to see much evidence of cloth dynamics. e.g (technically these gifs are from an earlier epoch but the _output_ barely changed after training for longer):

![RunWithTorch](https://github.com/hbertiche/NeuralClothSim/assets/30476356/1f1774d2-dc4c-48d9-b692-232952ee9275)

Compared to setting `motion=0.0`:

![RunWithTorchNoMotion](https://github.com/hbertiche/NeuralClothSim/assets/30476356/3b02ae03-b2fe-44a5-9e99-98aebdb52fbe)

Maybe it is just really subtle, but I can barely see a difference.

I think this is the same animation 1:26 in https://youtu.be/6HxXLBzRXFg?t=86 so there's clearly a pretty huge gap between the results I'm getting and what was obtained for that video.

On that 100th epoch, I got the following metrics:
```
m/Loss: 1.0557 - m/Stretch: 6.5842e-06 - m/Shear: 0.1285 - m/Bending: 0.1992 - m/Collision: 0.0041 - m/Gravity: 1.0049 - m/Inertia: 0.0049
```
FYI after just 5 _batches_ (not epochs!) of training, the metrics were:
```
m/Loss: 1.6685 - m/Stretch: 2.6247e-05 - m/Shear: 0.2460 - m/Bending: 0.0590 - m/Collision: 0.0359 - m/Gravity: 1.0188 - m/Inertia: 0.0046
```
So `Stretch`, `Shear`, `Bending`, `Collision` and `Gravity` (to a lesser extent) losses all improved quite a bit, but it looks like inertia loss barely changed (if anything, it got worse?). Perhaps inertia loss starting off so tiny could be the cause for the model seemingly not learning dynamics?

I also tried a very un-scientific test: printing the mean and max magnitude of values in the dynamic and static encodings (just before running the decoder):
```py
x_static_abs = tf.abs(x_static)
x_dynamic_abs = tf.abs(x_dynamic)
for i in range(tf_shape(x_static)[1]):
  x_static_slice = x_static_abs[:, i, :]
  x_dynamic_slice = x_dynamic_abs[:, i, :]
  x_static_mean = tf.math.reduce_mean(x_static_slice)
  x_static_max = tf.math.reduce_max(x_static_slice)
  x_dynamic_mean = tf.math.reduce_mean(x_dynamic_slice)
  x_dynamic_max = tf.math.reduce_max(x_dynamic_slice)
  print(f"Means: Static - {x_static_mean}, Dynamic - {x_dynamic_mean}\nMaximums: Static - {x_static_max}, Dynamic - {x_dynamic_max}")
```
and got (for the last frame on the above animation):
```
Means: Static - 0.015779726207256317, Dynamic - 0.013485459610819817
Maximums: Static - 0.14864373207092285, Dynamic - 0.062263891100883484
```
Obviously average/maximum magnitude across values in the encodings won't always entirely correlate with size of output deformations, but at least it looks like the dynamic encoder is having _some_ influence on the final output, just not anything that resembles coherent cloth dynamics.

My main question then: am I doing anything obviously wrong? I guess the most obvious thing left to try is to train for more epochs, but the paper did mention that simple garments should only take an hour to train (max a day). Training for 100 epochs took about a day on a 4070 and losses seem to be decreasing very slowly. Regardless, I will try running training over a few more days and update this issue if I managed to get a better result... Any other ideas for what I might be doing wrong (do I need to be training with a larger batch size?) or other assistance in general (perhaps sharing the exact config/set of Mixamo animations used to train the model shown in the paper?) would be very much appreciated. Thanks!!

**Comment by NathanielB123** [comment]
Follow up on this. I have now trained a model for 1000 epochs. The model has learnt to ignore dynamics completely:
```
Means: Static-0.010450625792145729 Dynamic-0.0
Maximums: Static-0.3061593174934387 Dynamic-0.0
```
I also tried printing out the number of zeroes in both the static and latent codes:
```
print(
  f"Zeroes: Static-{tf.reduce_sum(tf.cast(tf.equal(x_static_slice, 0), tf.int32))} Dynamic-{tf.reduce_sum(tf.cast(tf.equal(x_dynamic_slice, 0), tf.int32))}"
)
```
and got:
```
Zeroes: Static-457 Dynamic-512
```
So even the static latent code appears to be majority zeroes (both latent codes contain just 512 elements). i.e. perhaps the problems are not limited to the dynamic encoder; it seems like the model is just ending up with a ton of stuck ReLUs somehow (and it is especially bad w.r.t. dynamics because inertia loss is more unstable)?

I've experimented with various tweaks like freezing the weights of the dynamic/static encoder to train each individually, scaling up inertia loss, scaling up inertia loss over time, training with larger batch size (via training on CPU, I managed to get a batch size of 1024 running), training with smaller batch size, trying different temporal window sizes and motion augmentation percentages etc... and have not managed to achieve any results close to those shown in the paper and video...

As before, any ideas on what I might be doing wrong would be **hugely** appreciated!

**Comment by ChocoOreal** [comment]
any updates on the training @NathanielB123 ? From the demo video i think that having higher frame rates may improve the dynamics?

**Comment by NathanielB123** [comment]
> any updates on the training [@NathanielB123](https://github.com/NathanielB123) ? From the demo video i think that having higher frame rates may improve the dynamics?

I never managed to get results like the demo video unfortunately. I vaguely remember that with a lot of tuning it was possible to get the dynamic encoder to have a somewhat meaningful output, but it never looked nearly as impressive as that demo (especially looser garments like skirts always ended up looking broken). There is definitely a chance I could have just been doing something horribly wrong though!

**Comment by ChocoOreal** [comment]
> I never managed to get results like the demo video unfortunately. I vaguely remember that with a lot of tuning it was possible to get the dynamic encoder to have a somewhat meaningful output, but it never looked nearly as impressive as that demo (especially looser garments like skirts always ended up looking broken). There is definitely a chance I could have just been doing something horribly wrong though!

I see. How long did it take for you to complete the training?

**Comment by NathanielB123** [comment]
> I see. How long did it take for you to complete the training?

From what I remember, the model appears to converge pretty quickly (like after a few hours), but I did also train some checkpoints for up to a couple days just to see if that helped.

**Comment by ChocoOreal** [comment]
> From what I remember, the model appears to converge pretty quickly (like after a few hours), but I did also train some checkpoints for up to a couple days just to see if that helped.

I have just tried to train the model for a couple of epochs on the CMU sequence of AMASS dataset, the results did not turn out quite well. Here is what i got. Have you encountered this issue?

<img width="373" alt="Image" src="https://github.com/user-attachments/assets/b4b637df-5a0d-497b-80bb-af6d46de1f08" />

**Comment by NathanielB123** [comment]
> Have you encountered this issue?

Not exactly, though I used SMPL pose sequences. I did run into some similar-ish issues when garment and body rest poses didn't line up (i.e. T-pose vs A-pose) when trying to add more garments/bodies, but I think both the provided T-shirt `.obj` and SMPL body should both be in a resting T-pose.

**Comment by ChocoOreal** [comment]
> Not exactly, though I used SMPL pose sequences. I did run into some similar-ish issues when garment and body rest poses didn't line up (i.e. T-pose vs A-pose) when trying to add more garments/bodies, but I think both the provided T-shirt `.obj` and SMPL body should both be in a resting T-pose.

Weird issue honestly. I think what the model does is to predict the local deformation of garments, the rest should be left to skinning. And assuming that both the provided T-shirt and SMPL body are in the same rest pose, i dont think the misalignment between cloth and body poses should happen.

**Comment by NathanielB123** [comment]
> Weird issue honestly. I think what the model does is to predict the local deformation of garments, the rest should be left to skinning. And assuming that both the provided T-shirt and SMPL body are in the same rest pose, i dont think the misalignment between cloth and body poses should happen.

Yeah - you can pretty easily tweak the model to predict without deformations (i.e. view only the effect of skinning) by removing deformations here https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/model/ncs.py#L246 so that might be worth checking. Garment LBS parameters are derived from rest-pose geometry (https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/model/cloth.py#L64), so I think if skinning is somehow broken, the most likely culprit probably is rest poses being wrong

**Comment by ChocoOreal** [comment]
> Yeah - you can pretty easily tweak the model to predict without deformations (i.e. view only the effect of skinning) by removing deformations here https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/model/ncs.py#L246 so that might be worth checking. Garment LBS parameters are derived from rest-pose geometry (https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/model/cloth.py#L64), so I think if skinning is somehow broken, the most likely culprit probably is rest poses being wrong

I tried your suggestions and seems like skinning isnt the problem. Something probably went wrong with deformation prediction itself.

**Comment by ChocoOreal** [comment]
I notice in here https://github.com/hbertiche/NeuralClothSim/blob/237eb2c37ff65971cfcff111e95ef35933513ed8/ncs/model/ncs.py#L244 and https://github.com/hbertiche/NeuralClothSim/blob/237eb2c37ff65971cfcff111e95ef35933513ed8/ncs/model/ncs.py#L248, only the last frame in the body sequence while 3 frames in the garment sequence were used to compute the loss. I wonder why is there the mismatch here. Have you tried to look into this?

**Comment by NathanielB123** [comment]
> , only the last frame in the body sequence while 3 frames in the garment sequence were used to compute the loss. I wonder why is there the mismatch here. Have you tried to look into this?

I think this is intentional - note that when computing the "static" loss (collision, stretching, bending etc...), the last frame of garment vertices is taken as you would expect (https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/model/ncs.py#L126). 3 frames are necessary for the dynamic/inertia-based loss (which needs to penalise change in velocity) (https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/loss/losses.py#L177)

**Comment by ChocoOreal** [comment]
> I think this is intentional - note that when computing the "static" loss (collision, stretching, bending etc...), the last frame of garment vertices is taken as you would expect (https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/model/ncs.py#L126). 3 frames are necessary for the dynamic/inertia-based loss (which needs to penalise change in velocity) (https://github.com/hbertiche/NeuralClothSim/blob/main/ncs/loss/losses.py#L177)

Thanks for the insights, I also got the problem fixed by modifying this line https://github.com/hbertiche/NeuralClothSim/blob/237eb2c37ff65971cfcff111e95ef35933513ed8/ncs/utils/rotation.py#L103 to 
```dot = np.clip(dot, -1 + 1e-5, 1 - 1e-5)```. In the original code, dot can be 1, making omega equal 0 and then some "divided by 0" computation happened. It may be the main reason why I got weird issue. But still, the result is not so good just as you described.

**Comment by NathanielB123** [comment]
> modifying this line to `dot = np.clip(dot, -1 + 1e-5, 1 - 1e-5)`

Ahh yeah I think I fixed this very early when I started getting the divide by zero warnings. Makes sense that could cause some weirdness. As for other quality issues, yeah I don't think I can be much help. Nice to know I'm not alone in having difficulty with getting it working at least. If you ever do get some better results, please do say - I would love to know what I perhaps missed!

FYI, I personally had a lot more success in getting high-quality outputs from GNN-based approaches (such as https://github.com/Dolorousrtur/HOOD - there might be some even better models around now). The performance costs are unfortunately a lot higher of course.

**Comment by ChocoOreal** [comment]
> FYI, I personally had a lot more success in getting high-quality outputs from GNN-based approaches (such as https://github.com/Dolorousrtur/HOOD - there might be some even better models around now). The performance costs are unfortunately a lot higher of course.

I actually have tried inference on ContourCraft which is the upgraded version of HOOD by the same author. While the result is good, it is so computationally inefficient. It almost exhausts all of VRAM on my machine with their provided samples.
